### DISCUSSION TITLE
Typescript as embedded language

### DISCUSSION THREAD

**Discussion by ejfasting** [microsoft/vscode-discussions#1097] [Extension Development QnA] [answered]
Hi!

I'm working on an lsp server that supports embedded languages. There are examples on how to do it with html/css in the samples-repo, and i was wondering if the same logic could be used for typescript? In other words my lsp server will define codeblocks and ask the built-inn ts server to give back completionitems etc. 

I also need to pass a *.d.ts with som additional types, but im unsure how i would do that from my lsp server. In monaco you can use addExtraLib, so would npmLocation be the equivalent? 

Any input/examples are much appreciated!

**Comment by tamuratak** [comment, marked as answer]
You would be interested in Volar.js

> Volar.js is architected to support any file format that involves embedded languages - not just Vue, but also Astro, Svelte, or even Angular. It is also capable of implementing regular single-language LSP servers such as TypeScript, CSS, and HTML.
> https://volarjs.dev/blog/volar-a-new-beginning/

See https://github.com/volarjs/starter

**Reply by ejfasting** [reply]
This looks very promising, thank you!

I havent gotten as far as testing how to 'inject' custom types into the tsserver. I assume this is done by creating a Ts Plugin: https://code.visualstudio.com/api/references/contribution-points#Plugin-configuration 
The question is if a ts plugin can be used on the server-side, or if it HAS to be use in the extension/client? Haven't found any good examples on how to do it from the server, unfortunately.
